src/main/java/slp/core/modeling/runners/ModelRunner.java

```diff
@@ -0,0 +1,489 @@
+package slp.core.modeling.runners;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.DoubleSummaryStatistics;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import slp.core.lexing.runners.LexerRunner;
+import slp.core.modeling.Model;
+import slp.core.modeling.ngram.NGramModel;
+import slp.core.translating.Vocabulary;
+import slp.core.util.Pair;
+
+/**
+ * The this class provides the third step in modeling (after lexing, translating).
+ * It requires a {@linkplain LexerRunner} and {@linkplain Model} and exposes modeling methods for convenience.
+ * 
+ * @author Vincent Hellendoorn
+ *
+ */
+public class ModelRunner {
+	
+	private static final double INV_NEG_LOG_2 = -1.0/Math.log(2);
+	public static final int DEFAULT_NGRAM_ORDER = 6;
+	
+	public static int GLOBAL_PREDICTION_CUTOFF = 10;
+	
+	protected final LexerRunner lexerRunner;
+	protected final Model model;
+	
+	private boolean selfTesting = false;
+	private Nester nester;
+	
+	public ModelRunner(LexerRunner lexerRunner, Model model) {
+		this.lexerRunner = lexerRunner;
+		this.model = model;
+	}
+	
+	public LexerRunner getLexerRunner() {
+		return this.lexerRunner;
+	}
+	
+	public Model getModel() {
+		return this.model;
+	}
+	
+	public static int getPredictionCutoff() {
+		return GLOBAL_PREDICTION_CUTOFF;
+	}
+
+	public static void setPredictionCutoff(int cutoff) {
+		GLOBAL_PREDICTION_CUTOFF = cutoff;
+	}
+
+	public Vocabulary getVocabulary() {
+		return this.lexerRunner.getVocabulary();
+	}
+	
+	/**
+	 * Indicate that we are testing on the training set,
+	 * which means we must 'forget' any files prior to modeling them and re-learn them afterwards.
+	 * <br/>
+	 * If nesting is enabled, this may reconfigure the nesting setup to match.
+	 * For better performance, set self-testing before enabling nesting.
+	 */
+	public void setSelfTesting(boolean selfTesting) {
+		if (selfTesting != this.selfTesting) {
+			this.selfTesting = selfTesting;
+			// If self-testing has changed and this model is already nested,
+			// we must re-calibrate the nester as well
+			if (this.isNested()) this.setNested(this.nester.getTestRoot());
+		}
+	}
+	
+	/**
+	 * Returns whether or not the model is set up to run self-testing (training on test-set)
+	 */
+	public boolean isSelfTesting() {
+		return this.selfTesting;
+	}
+	
+	/**
+	 * <b>Important:</b> if self-testing is the goal, please enable it before enabling nested mode!
+	 * Otherwise, self-testing will learn a model over all test data now.
```

Comment:
In general I'm wary of "important" notes in comments that aren't enforced in code. Is this the kind of constraint that could be integrated directly into the workflow? For example one way to do this might be by making "self-test" only accessible through the `.setNested` call. That way it'd be impossible to perform these operations in the wrong order.

Reply:
Yeah I really didn't like that part either; it was a pain just to get it to behave correctly regardless of the order in which those two (`setSelfTest` & `setNested`) are invoked. I was thinking of making a method called `configure`, in which both fields must be set at once, which sounds a bit like what you are suggesting. The downside is that such a method must ask for a `testRoot` File. Perhaps an overloaded version with and without that parameter would make sense?